Mathlib/Algebra/Order/Incidence.lean

```diff
@@ -0,0 +1,645 @@
+
+namespace IncidenceAlgebra
+section Zero
+variable [Zero 𝕜] [LE α] {a b : α}
+
+instance instFunLike : FunLike (IncidenceAlgebra 𝕜 α) α (α → 𝕜) where
+  coe := toFun
+  coe_injective' f g h := by cases f; cases g; congr
+
+lemma apply_eq_zero_of_not_le (h : ¬a ≤ b) (f : IncidenceAlgebra 𝕜 α) : f a b = 0 :=
+  eq_zero_of_not_le' _ h
+
+lemma le_of_ne_zero {f : IncidenceAlgebra 𝕜 α} : f a b ≠ 0 → a ≤ b :=
+  not_imp_comm.1 fun h ↦ apply_eq_zero_of_not_le h _
+
+section Coes
+
+-- this must come after the `FunLike` instance
+initialize_simps_projections IncidenceAlgebra (toFun → apply)
+
+@[simp] lemma toFun_eq_coe (f : IncidenceAlgebra 𝕜 α) : f.toFun = f := rfl
+@[simp, norm_cast] lemma coe_mk (f : α → α → 𝕜) (h) : (mk f h : α → α → 𝕜) = f := rfl
+
+lemma coe_inj {f g : IncidenceAlgebra 𝕜 α} : (f : α → α → 𝕜) = g ↔ f = g :=
+  DFunLike.coe_injective.eq_iff
+
+@[ext]
+lemma ext ⦃f g : IncidenceAlgebra 𝕜 α⦄ (h : ∀ a b, a ≤ b → f a b = g a b) : f = g := by
+  refine DFunLike.coe_injective' (funext₂ fun a b ↦ ?_)
+  by_cases hab : a ≤ b
+  · exact h _ _ hab
+  · rw [apply_eq_zero_of_not_le hab, apply_eq_zero_of_not_le hab]
+
+@[simp] lemma mk_coe (f : IncidenceAlgebra 𝕜 α) (h) : mk f h = f := rfl
+
+end Coes
+
+/-! ### Additive and multiplicative structure -/
+
+instance instZero : Zero (IncidenceAlgebra 𝕜 α) := ⟨⟨fun _ _ ↦ 0, fun _ _ _ ↦ rfl⟩⟩
+instance instInhabited : Inhabited (IncidenceAlgebra 𝕜 α) := ⟨0⟩
+
+@[simp, norm_cast] lemma coe_zero : ⇑(0 : IncidenceAlgebra 𝕜 α) = 0 := rfl
+lemma zero_apply (a b : α) : (0 : IncidenceAlgebra 𝕜 α) a b = 0 := rfl
+
+end Zero
+
+section Add
+variable [AddZeroClass 𝕜] [LE α]
+
+instance instAdd : Add (IncidenceAlgebra 𝕜 α) where
+  add f g := ⟨f + g, fun a b h ↦ by simp_rw [Pi.add_apply, apply_eq_zero_of_not_le h, zero_add]⟩
+
+@[simp, norm_cast] lemma coe_add (f g : IncidenceAlgebra 𝕜 α) : ⇑(f + g) = f + g := rfl
+lemma add_apply (f g : IncidenceAlgebra 𝕜 α) (a b : α) : (f + g) a b = f a b + g a b := rfl
+
+end Add
+
+section Smul
+variable {M : Type*} [Zero 𝕜] [LE α] [SMulZeroClass M 𝕜]
+
+instance instSmulZeroClassRight : SMulZeroClass M (IncidenceAlgebra 𝕜 α) where
+  smul c f :=
+    ⟨c • ⇑f, fun a b hab ↦ by simp_rw [Pi.smul_apply, apply_eq_zero_of_not_le hab, smul_zero]⟩
+  smul_zero c := by ext; exact smul_zero _
+
+@[simp, norm_cast] lemma coe_constSMul (c : M) (f : IncidenceAlgebra 𝕜 α) : ⇑(c • f) = c • ⇑f := rfl
+
+lemma constSMul_apply (c : M) (f : IncidenceAlgebra 𝕜 α) (a b : α) : (c • f) a b = c • f a b := rfl
+
+end Smul
+
+instance instAddMonoid [AddMonoid 𝕜] [LE α] : AddMonoid (IncidenceAlgebra 𝕜 α) :=
+  DFunLike.coe_injective.addMonoid _ coe_zero coe_add fun _ _ ↦ rfl
+
+instance instAddCommMonoid [AddCommMonoid 𝕜] [LE α] : AddCommMonoid (IncidenceAlgebra 𝕜 α) :=
+  DFunLike.coe_injective.addCommMonoid _ coe_zero coe_add fun _ _ ↦ rfl
+
+section AddGroup
+variable [AddGroup 𝕜] [LE α]
+
+instance instNeg : Neg (IncidenceAlgebra 𝕜 α) where
```

Comment:
Should work with weaker assumptions? Perhaps similar below

Reply:
I need `neg_zero`. I don't think I want to be reasoning about `neg_zero` in this file